### ISSUE TITLE
DataProvider type should be more strict

### ISSUE THREAD

**Issue by etienne-bondot** [marmelab/react-admin#6437]
**Is your feature request related to a problem? Please describe.**
When I want to enhance my custom dataProvider, I want the dataProvider type to be strict and correct.
For now, as the dataProvider type is kind of lazy because of the 

```
export type DataProvider = {
    ...

    [key: string]: any;
};
```

It could contained barely anything. 

So when I added a custom method to my dataProvider, even when the type is properly extended, I get

![image](https://user-images.githubusercontent.com/2827929/125750354-65242129-142c-4da1-9858-28b2458a3504.png)

**Describe the solution you'd like**
I suggest to remove this last key and so I can strictly extend the default type like 

```
interface ExtendedDataProvider extends DataProvider {
  getStatics: <RecordType extends Record = Record>(
    resource: string,
    filter?: { [key: string]: string }
  ) => Promise<GetStaticsResult<RecordType>>;
}
```

Did I do something wrong or is there any workaround to get the proper method type extending the DataProvider type ?

**Comment by djhi** [comment]
Yes, I haven't found the time to do it yet but `useDataProvider` should accept and return a generic type which must extend `DataProvider`

**Comment by WiXSL** [comment]
This has been implemented in #6605.
So, I'm closing this issue.
